### ISSUE TITLE
tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#127232]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13.5

### Current behavior?

`tf.keras.ops.numpy.logaddexp` returns an incorrect first derivative at equal inputs.

The reproducer evaluates `tf.keras.ops.numpy.logaddexp(t, [-4, 0, 5])` inside a weighted scalar expression at `t = -4.0`. In the first term, the two inputs to `logaddexp` are equal. Since `log(exp(x) + exp(y))` is smooth at equality, the partial derivative with respect to the first input should be `1/2` at that point. A stabilization branch should preserve this derivative.

The forward value is computed correctly as `-5.770614664699297`, but reverse-mode autodiff returns `1.0359107226361899` for the derivative with respect to `t`. The expected derivative is approximately `0.53591072263619`.

Expected behavior?

The first derivative of the weighted `tf.keras.ops.numpy.logaddexp` expression at `t = -4.0` should be close to `0.53591072263619`, not `1.0359107226361899`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    y = tf.keras.ops.numpy.logaddexp(
        t,
        tf.constant([-4, 0, 5], dtype=t.dtype),
    )
    return y[0] + 2 * y[1] - y[2] / 2

x = tf.constant(-4.0, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.53591072263619

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: -5.770614664699297
actual: 1.0359107226361899
expected: 0.53591072263619
BUG REPRODUCED: tf.keras.ops.numpy.logaddexp returns incorrect gradient at equal inputs
```

**Comment by jdymitarai** [comment]
I have identified the root cause and submitted a fix in **keras-team/keras#23616**.

### Root Cause Analysis
In `keras/src/backend/tensorflow/numpy.py`, `logaddexp` and `logaddexp2` used:
```python
amax = tf.maximum(x1, x2)
...
return tf.where(
    tf.math.is_nan(delta),
    x1 + x2,
    amax + tf.math.log1p(tf.math.exp(-tf.abs(delta))),
)
```
When $x_1 = x_2$:
1. `tf.maximum(x1, x2)` routes 100% of the upstream gradient to $x_1$ and 0% to $x_2$ because $x_1 \ge x_2$ is true.
2. The subgradient of `tf.abs(delta)` at $\delta = 0$ is $0$, which causes the second term `log1p(exp(-abs(delta)))` to produce zero gradient.
3. Consequently, reverse-mode autodiff at equal inputs collapsed to `(1.0, 0.0)` instead of the correct mathematical gradient `(0.5, 0.5)`. In your reproducer, this caused the partial derivative of `y[0]` with respect to `t` to evaluate to `1.0` rather than `0.5`, explaining the exact `0.5` discrepancy in the scalar output (`1.03591...` vs `0.53591...`).

### Resolution
We reformulate the numerical stabilization using `safe_max = tf.stop_gradient(...)`, matching standard LogSumExp implementations:
```python
amax = tf.maximum(x1, x2)
safe_max = tf.stop_gradient(
    tf.where(tf.math.is_finite(amax), amax, tf.zeros_like(amax))
)
exp_1 = tf.math.exp(x1 - safe_max)
exp_2 = tf.math.exp(x2 - safe_max)
return tf.where(
    tf.math.is_nan(delta),
    x1 + x2,
    safe_max + tf.math.log(exp_1 + exp_2),
)
```
Because `safe_max` is stopped from backpropagation, reverse-mode autodiff evaluates:
$$\frac{\partial}{\partial x_1} \left( S + \ln(e^{x_1 - S} + e^{x_2 - S}) \right) = \frac{e^{x_1 - S}}{e^{x_1 - S} + e^{x_2 - S}} = \text{softmax}_1$$
At $x_1 = x_2$, both partial derivatives are strictly $0.5$. The identical stabilization is applied to `logaddexp2`.

Fixed in https://github.com/keras-team/keras/pull/23616 with comprehensive unit tests across TensorFlow, JAX, and PyTorch backends, including the exact weighted expression from this issue.
